File: codeiq/tyama_codeiq711_iter.vb
imports System
imports System.Collections.Generic
module CodeIQRoute
	iterator function permutation(of T as IComparable)(ByVal x as List(of T),optional ByVal _n as integer?=Nothing) as IEnumerable(of List(of T))
		dim n as integer=if(_n.HasValue,_n,x.Count)
		if n<0 orelse x.Count<n
			return
		end If
		dim a as List(of T)=new List(of T)(x)
		a.Sort()
		dim i as integer
		yield a.GetRange(0,n)
		while true
			a.Reverse(n,a.Count-n)
			for i=a.Count-2 to 0 step -1
				if a(i).CompareTo(a(i+1))<0
					exit for
				end if
			next
			if i<0
				a.Reverse(0,a.Count)
				return
			end if
			dim k as integer=i
			for i=a.Count-1 to k+1 step -1
				if a(k).CompareTo(a(i))<0
					exit for
				end if
			next
			dim l as integer=i
			dim z as T=a(k):a(k)=a(l):a(l)=z
			a.Reverse(k+1,a.Count-(k+1))
			yield a.GetRange(0,n)
		end while
	end function

	const N=6
	sub Main(ByVal args() as String)
		dim r,i as integer
		dim e0 as List(of integer)=new List(of integer)()
		dim f0 as List(of integer)=new List(of integer)()
		for i=0 to N-1
			e0.Add(0)
			f0.Add(0)
		next
		for i=0 to N-1
			e0.Add(1)
			f0.Add(1)
		next
		for each _e as List(of integer) in permutation(e0)
			for each _f as List(of integer) in permutation(f0)
				dim flg as integer=0
				dim zero1 as int=0
				dim zero2 as integer=N
				dim one1 as integer=0
				dim one2 as integer=N
				for i=0 to N*2-1
					if _e(i)==0
						zero1+=1
					end if
					if _e(i)==1
						one1+=1
					end if
					if _f(N*2-1-i)==0
						zero2-=1
					end if
					if _f(N*2-1-i)==1
						one2-=1
					end if
					if zero1==zero2
						flg+=1
					end if
					if one1==one2
						flg+=1
					end if
				next
				if flg>=2
					r+=1
				end if
			next
		next
		Console.WriteLine(r)
	end sub
end module

File: codeiq/tyama_codeiq711_iter.cs
using System;
using System.Collections.Generic;
class CodeIQRoute{
	static IEnumerable<List<T>> Permutation<T>(List<T> x,int? _n=null) where T : IComparable<T>{
		int n=_n ?? x.Count;
		if(n<0||x.Count<n)yield break;
		List<T> a=new List<T>(x);
		a.Sort();
		yield return a.GetRange(0,n);
		for(;;){
			int i;
			a.Reverse(n,a.Count-n);
			for(i=a.Count-2;i>=0;i--)if(a[i].CompareTo(a[i+1])<0)break;
			if(i<0){
				//a.Reverse(0,a.Count);
				/*yield*/ break;
			}
			int k=i;
			for(i=a.Count-1;i>=k+1;i--)if(a[k].CompareTo(a[i])<0)break;
			int l=i;
			T z=a[k];a[k]=a[l];a[l]=z;
			a.Reverse(k+1,a.Count-(k+1));
			yield return a.GetRange(0,n);
		}
	}
	const int N=6;
	static void Main(){
		int r=0,i;
		List<int>e0=new List<int>(),f0=new List<int>();
		for(i=0;i<N;i++){e0.Add(0);f0.Add(0);}
		for(i=0;i<N;i++){e0.Add(1);f0.Add(1);}
		foreach(List<int> _e in Permutation(e0)){
			foreach(List<int> _f in Permutation(f0)){
				int flg=0;
				int zero1=0;
				int zero2=N;
				int one1=0;
				int one2=N;
				for(i=0;i<N*2;i++){
					if(_e[i]==0)zero1++;
					if(_e[i]==1)one1++;
					if(_f[N*2-1-i]==0)zero2--;
					if(_f[N*2-1-i]==1)one2--;
					if(zero1==zero2)flg++;
					if(one1==one2)flg++;
				}
				if(flg>=2)r++;
			}
		}
		Console.WriteLine(r);
	}
}